subprojects/core/src/main/java/org/gradle/internal/fingerprint/impl/DefaultFileCollectionSnapshotter.java

```diff
@@ -70,21 +68,20 @@ public void visitGenericFileTree(FileTreeInternal fileTree) {
 
         @Override
         public void visitFileTree(File root, PatternSet patterns, FileTreeInternal fileTree) {
-            MerkleDirectorySnapshotBuilder builder = MerkleDirectorySnapshotBuilder.sortingRequired();
-            virtualFileSystem.read(root.getAbsolutePath(), new PatternSetSnapshottingFilter(patterns, stat), builder);
-            FileSystemLocationSnapshot result = builder.getResult();
-            roots.add((result == null || result.getType() == FileType.Missing) ? FileSystemSnapshot.EMPTY : result);
+            virtualFileSystem.read(
+                root.getAbsolutePath(),
+                new PatternSetSnapshottingFilter(patterns, stat),
```

Comment:
Should we store the `PatternSetSnapshottingFilter` in place of `patterns`? If not, I think we should use a factory to skip creating an instance for empty an `patterns`.

Reply:
The class is very light-weight and probably doesn't make a difference.